bt_ddos_shield/miner_shield.py

```diff
@@ -186,85 +208,120 @@ def _handle_validators_changed(self):
+            f", new_validators count={len(new_validators)}, changed_validators count={len(changed_validators)}"
         )
 
         for validator in deprecated_validators:
-            self.event_processor.add_event(MinerShieldEvent(f"Removing validator {validator}"))
+            self._event(f"Removing validator {validator}")
 
             if validator in current_state.active_addresses:
                 self.address_manager.remove_address(current_state.active_addresses[validator])
 
             self.state_manager.remove_validator(validator)
 
-        # TODO handle new_validators and changed_validators
+        for validator in new_validators:
+            self._event(f"Adding validator {validator}")
+
+            new_address = self.address_manager.create_address()
+
+            try:
+                self.state_manager.add_validator(validator, fetched_validators[validator], new_address)
+            except Exception as e:
+                self.address_manager.remove_address(new_address.address_id)
+                raise e
 
-        if deprecated_validators or new_validators or changed_validators:
+        for validator, new_key in changed_validators:
+            self._event(f"Updating validator {validator}")
+            self.state_manager.update_validator(validator, new_key)
+
+        if force_update_manifest or deprecated_validators or new_validators or changed_validators:
             # if anything changed update manifest file and publish new version to blockchain
-            # TODO also check state of shield if manifest was published at all
-            pass
+            self._add_task(MinerShieldUpdateManifestTask())
 
     def _handle_ban_validator(self, validator_hotkey: Hotkey):
         """
-        Ban validator by its hotkey. It will update manifest file and publish info about new file version to blockchain.
+        Ban validator by its hotkey. If something changed, MinerShieldValidatorsChangedTask will apply asynchronously
+        this change where needed.
         """
-        # TODO
-        pass
+        self.state_manager.ban_validator(validator_hotkey)
+        self._event(f"Validator {validator_hotkey} added to banned set")
+        self._add_task(MinerShieldValidatorsChangedTask())
 
+    def _handle_update_manifest(self):
+        """
+        Update manifest file and schedule publishing it to blockchain.
+        """
+        address = self.manifest_manager.add_mapping_file(self.state_manager.get_state().active_addresses)
+        self._event(f"Manifest updated, new address: {address}")
+        self._add_task(MinerShieldPublishManifestTask(address))
 
-class MinerShieldTask:
+    def _handle_publish_manifest(self, address: Address):
+        """
+        Publish info about current manifest file to blockchain.
+        """
+        current_address = Address.deserialize(self.blockchain_manager.get(self.miner_hotkey))
```

Comment:
```suggestion
        current_address = self.ADDRESS_CLASS.deserialize(self.blockchain_manager.get(self.miner_hotkey))
```
```py
class costam:
    ADDRESS_CLASS: type = Address

class MyCostam:
    ADDRESS_CLASS: type = MyAddress

```

Reply:
`ClassVar[type[Address]]` though?

Reply:
I've done it other, better way with usage of Serializer class